posix: add logging to pthread, pthread_mutex, and pthread_cond

cfriedt opened the pull request:
For A Long Time™️ there was zero logging in our pthread implementation. This made it sometimes difficult to know what was happening under the hood with the POSIX library.

Here, we add error logging to the primary paths of interest in POSIX threads - namely to `pthread_t`, `pthread_mutex_t`, and `pthread_cond_t` operations.

Debug-level logging is also implemented when more verbose output is desired.

Logging is also available by default for any of the pooled resources via the Kconfig template, although it has not yet been implemented for less frequently-traveled types like `pthread_barrier_t`, `pthread_rwlock_t`, and `pthread_spinlock_t`. Those can be implemented on an as-needed basis.

For implementing various APIs on top of pthreads, such as ISO C, and ISO C++ threads, this kind of logging is quite helpful.

Commit:
posix: add logging to pooled resource templates

Support logging for all POSIX pooled resource types such as
pthread_t, pthread_mutex_t, and pthread_cond_t.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: pthread: add logging for POSIX threads

Add logging to POSIX threads.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: mutex: add logging for POSIX mutexes

Add logging for POSIX mutex operations.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

Commit:
posix: cond: add logging for condition variables

Add logging for POSIX condition variable operations.

Signed-off-by: Christopher Friedt <cfriedt@meta.com>

ycsin commented on lib/posix/pthread.c:
```diff
@@ -662,7 +694,10 @@ int pthread_join(pthread_t pthread, void **status)
 	/* other possibilities? */
 	__ASSERT_NO_MSG(err == 0);
 
+	LOG_DBG("Joined  pthread %x", pthread);
```
```suggestion
	LOG_DBG("Joined pthread %x", pthread);
```

cfriedt replied:
That spacing is kind of intentional to get related messages to align vertically. I can remove if necessary though.

ycsin commented on lib/posix/mutex.c:
```diff
@@ -148,18 +160,24 @@ static int acquire_mutex(pthread_mutex_t *mu, k_timeout_t timeout)
 	}
 
+	if (ret == 0) {
+		LOG_DBG("Locked mutex  %x", *mu);
```
```suggestion
		LOG_DBG("Locked mutex %x", *mu);
```

cfriedt replied:
Same as https://github.com/zephyrproject-rtos/zephyr/pull/62771#discussion_r1328695189

ycsin replied:
ops

ycsin reviewed (dismissed):
just nits

cfriedt commented:
* removed extra spaces from strings (were there for vertical alignment, but I think those are removed anyway)
* added a few debug logs to `cond.c` for more verbose messaging of wait, signal, and broadcast operations

cfriedt commented:
* log pointer value for actual k_thread, k_mutex, or k_condvar when possible
